### PR TITLE
argSUSed: lib (so also the bit of module that is built for lib), include

### PR DESCRIPTION
**Pull request by nabijaczleweli**
### Motivation and Context
There are **2** unused arguments among us.

### Description
~~Draft because on top of #12835 and #12067, otherwise should be~~ g2g except as indicated below.

The main revelation over my previous attempt at this was making ASSERTs et al do (void)sizeof(args)...: this doesn't ODR-use the arguments, so they get elided as usual, but it marks them as used (and actually verifies that they compile, which is a nice bonus), bringing the total diffstat to just
> 113 files changed, 446 insertions(+), 331 deletions(-)

across 97 commits, which is *more* than fine, imo
These are sequestered to the first two commits.

Commits that start with a `!` are more than purely editorial. In zfs sendrecv, I think this was a bug that I fixed. In arc.c I removed the mutex argument, which the comment says is used but isn't, and is only used by the caller?

In rdwr_efi.c, do we want efi_type() anymore? All it is right now is some #if 0ed Solaris code that returns ENOSYS, and is only applicable for the Solaris blockdev layer. In the Illumos gate, there's a single user: rmformat(1); I recommend a read of the manual as a blast from the past, but, well.

### How Has This Been Tested?
`make -C lib/` builds clean on clang 14 when configured with `-Wall -Wextra -Wno-sign-compare -Wno-missing-field-initializers` in CFLAGS

`git -C lib grep -i argsused` and the equivalent invocations for `include` and `module/{avl,icp,lua,nvpair,spl,unicode,zcommon,zstd}` are empty. The keen-eyed observer will note that this leaves only `module/zfs` and `module/os` (f/s, f/z, and l/z). (Well, and `cstyle.pl`.)

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes. – none should apply
- [ ] I have run the ZFS Test Suite with this change applied. – CI take my hand
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.co/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by). – needs mass replacing on `;` when ready

### CODE REVIEW

**Comment by nabijaczleweli** [comment]
Rebased with full commit messages, should be g2g, but see questions in OP

File: module/os/freebsd/zfs/zfs_vfsops.c [hunk 1]
```diff
@@ -1806,14 +1806,10 @@ zfs_fhtovp(vfs_t *vfsp, fid_t *fidp, int flags, vnode_t **vpp)
 	if (zfsvfs->z_parent == zfsvfs && fidp->fid_len == LONG_FID_LEN) {
 		zfid_long_t	*zlfid = (zfid_long_t *)fidp;
 		uint64_t	objsetid = 0;
-		uint64_t	setgen = 0;
 
 		for (i = 0; i < sizeof (zlfid->zf_setid); i++)
 			objsetid |= ((uint64_t)zlfid->zf_setid[i]) << (8 * i);
 
-		for (i = 0; i < sizeof (zlfid->zf_setgen); i++)
-			setgen |= ((uint64_t)zlfid->zf_setgen[i]) << (8 * i);
```
Comment (behlendorf) [inline module/os/freebsd/zfs/zfs_vfsops.c:1815]: There's also a good chance this is bug, the Linux counterpart to this function `zfs_vget` does make use of `setgen` for an error case.  This would also be a good candidate to move to it's own PR so we can get the FreeBSD folks to take a closer look.

File: module/zfs/arc.c [hunk 2]
```diff
@@ -1947,7 +1948,7 @@ arc_fill_hdr_crypt(arc_buf_hdr_t *hdr, kmutex_t *hash_lock, spa_t *spa,
  * arc_buf_fill().
  */
 static void
-arc_buf_untransform_in_place(arc_buf_t *buf, kmutex_t *hash_lock)
+arc_buf_untransform_in_place(arc_buf_t *buf)
```
Comment (behlendorf) [inline module/zfs/arc.c:1951]: Just for context the `hash_lock` still actually may be checked here through the `HDR_EMPTY_OR_LOCKED()` check.  It doesn't need so this looks fine.

File: lib/libzutil/os/freebsd/zutil_import_os.c [hunk 3]
```diff
@@ -249,6 +249,6 @@ zfs_dev_flush(int fd __unused)
 }
 
 void
-update_vdevs_config_dev_sysfs_path(nvlist_t *config)
+update_vdevs_config_dev_sysfs_path(nvlist_t *config __unused)
```
Comment (behlendorf) [inline lib/libzutil/os/freebsd/zutil_import_os.c:252]: Why `__unused` here and not `(void) config`?

File: module/zfs/dnode.c [hunk 4]
```diff
@@ -1621,7 +1621,9 @@ dnode_rele_and_unlock(dnode_t *dn, void *tag, boolean_t evicting)
 	 * other direct or indirect hold on the dnode must first drop the dnode
 	 * handle.
 	 */
+#ifdef ZFS_DEBUG
 	ASSERT(refs > 0 || dnh->dnh_zrlock.zr_owner != curthread);
+#endif
```
Comment (behlendorf) [inline module/zfs/dnode.c:1623]: Why was this needed? Are the new ASSERT definitions not happy with this when !ZFS_DEBUG?

**Review by behlendorf — COMMENTED** [review]
Generally looks good.  In order to move this forward what I'd like to do is move the commits which include functional changes, or I commented on specifically, in to their own PRs.  We'll want to round up additional reviewers for those.

For the rest it would be nice to squash the patch stack a bit.  One option would be merging similar commits together by library/module or header.

File: module/zfs/dnode.c (hunk 4 above)
Comment (nabijaczleweli) [reply module/zfs/dnode.c:1623]: Yes, `zr_owner` only exists when ZFS_DEBUG. This is the only instance of the new asserts requiring this, so I decided against making an ASSERT_SWALLOW.

File: module/os/freebsd/zfs/vdev_file.c [hunk 5]
```diff
@@ -227,11 +226,11 @@ vdev_file_io_strategy(void *arg)
 	ASSERT(zio->io_type == ZIO_TYPE_READ || zio->io_type == ZIO_TYPE_WRITE);
 	if (zio->io_type == ZIO_TYPE_READ) {
 		buf = abd_borrow_buf(zio->io_abd, zio->io_size);
-		err = zfs_file_pread(vf->vf_file, buf, size, off, &resid);
+		(void) zfs_file_pread(vf->vf_file, buf, size, off, &resid);
 		abd_return_buf_copy(zio->io_abd, buf, size);
 	} else {
 		buf = abd_borrow_buf_copy(zio->io_abd, zio->io_size);
-		err = zfs_file_pwrite(vf->vf_file, buf, size, off, &resid);
+		(void) zfs_file_pwrite(vf->vf_file, buf, size, off, &resid);
```
[the remark this answers is not in the document]
Comment (nabijaczleweli) [reply module/os/freebsd/zfs/vdev_file.c:233]: #12904

File: module/os/freebsd/zfs/zfs_vfsops.c (hunk 1 above)
Comment (nabijaczleweli) [reply module/os/freebsd/zfs/zfs_vfsops.c:1815]: #12905

File: lib/libzutil/os/freebsd/zutil_import_os.c (hunk 3 above)
Comment (nabijaczleweli) [reply lib/libzutil/os/freebsd/zutil_import_os.c:252]: Because there's __unused above, too, but fair point, I've changed them both to void casts.

**Comment by nabijaczleweli** [comment]
Updated. Moved the remaining one (libzfs/sendrecv) to #12906, others as noted.

As it stands, here's all the sus commits, sorted by amount of commits per section:
```
include: dmu.h
libefi: efi_type
libzfs_core
libshare: nfs: is_shared_cb
libshare: platform: nfs: disable_share_impl
libzpool: util
libzpool: kernel
libzpool: taskq
libzutil: import: find_import_impl
libzutil: import: platform
libzutil: import: platform: is_mpath_whole_disk
libnvpair: alloc_fixed: free
libnvpair: main: print_list_*
libnvpair: alloc_system: free
libnvpair: fnvpair: free
libnvpair: nvpair
libzfs: util: linux
libzfs: changelist
libzfs: config: node_compare
libzfs: crypto
libzfs: dataset: userquota_propname_decode
libzfs: mountpoint
libzfs: status
libzfs: mount: platforms
libzfs: fletcher
libzfs: iter
module: avl
module: zcommon: zprop
module: zfs: linux: abd
module: zfs: linux: arc
module: zfs: debug: linux
module: zfs: freebsd: znode
module: zfs: freebsd: spa_os
module: zfs: bpobj
module: zfs: bptree
module: zfs: dmu
module: zfs: dmu: diff
module: zfs: dmu: objset
module: zfs: dmu: redact
module: zfs: dmu: send
module: zfs: dmu: traverse
module: zfs: dmu: zfetch
module: zfs: dnode
module: zfs: dsl: bookmark
module: zfs: dsl: dataset
module: zfs: dsl: destroy
module: zfs: dsl: dir
module: zfs: dsl: pool
module: zfs: dsl: prop
module: zfs: dsl: scan
module: zfs: dsl: synctask: null_checkfunc
module: zfs: fm
module: zfs: gzip
module: zfs: lz4
module: zfs: lzjb
module: zfs: metaslab
module: zfs: range_tree
module: zfs: sa
module: zfs: sha256
module: zfs: spa
module: zfs: spa: checkpoint
module: zfs: spa: history
module: zfs: spa: misc
module: zfs: spa: vdev
module: zfs: vdev: draid
module: zfs: vdev: file: platforms
module: zfs: vdev: indirect
module: zfs: vdev: initialize
module: zfs: vdev: mirror
module: zfs: vdev: missing
module: zfs: vdev: raidz
module: zfs: vdev: removal
module: zfs: vdev: trim
module: zfs: zcp
module: zfs: zcp: synctask
module: zfs: racct: platforms
module: zfs: zil
module: zfs: zio
module: zfs: zio: checksum
module: zfs: zio: compress
module: zfs: zio: crypt: platforms
module: zfs: abd
module: zfs: arc
module: zfs: dbuf
module: zfs: zfs_fm
module: icp: algs: gcm
module: icp: algs: ccm
module: icp: io: aes
module: icp: io: sha1_mod
module: icp: core: kcf: mech_tabs
module: icp: core: kcf: prov_lib
module: icp: io: sha2_mod
module: icp: os: modhash
module: icp: os: modconf
module: icp: core: kcf: sched
module: icp: io: skein_mod
module: icp: remaining
```

Here's variant lite (libraries grouped, modules grouped by the subsystem):
```
include: dmu.h
libefi: efi_type
libzfs_core

libshare: nfs: is_shared_cb

libzpool

libzutil: import

libnvpair

libzfs

module: avl
module: zcommon: zprop

module: zfs: linux
module: zfs: freebsd

module: zfs: bpobj
module: zfs: bptree
module: zfs: dmu
module: zfs: dnode
module: zfs: dsl
module: zfs: fm
module: zfs: gzip
module: zfs: lz4
module: zfs: lzjb
module: zfs: metaslab
module: zfs: range_tree
module: zfs: sa
module: zfs: sha256
module: zfs: spa
module: zfs: vdev
module: zfs: zcp
module: zfs: racct: platforms
module: zfs: zil
module: zfs: zio
module: zfs: abd
module: zfs: arc
module: zfs: dbuf
module: zfs: zfs_fm
module: icp: algs
module: icp: io
module: icp: core: kcf
module: icp: os: modhash
module: icp: os: modconf
module: icp: remaining
```

Or variant heavy (grouped by module, I kept the linux/freebsd bits separate because I usually do but in variant heavy+, let's say, these can also meld together):
```
include: dmu.h
libefi: efi_type
libzfs_core

libshare: nfs: is_shared_cb

libzpool

libzutil: import

libnvpair

libzfs

module: avl
module: zcommon: zprop

module: zfs: linux
module: zfs: freebsd
module: zfs

module: icp
```

Which one strikes you best?

**Comment by behlendorf** [comment]
Let's go with the last variant, that seems like a sufficient level of granularity for this kind of change.  The `module: zfs` commit I imagine with be substantially larger than the others but that's not a problem.

```
include: dmu.h
libefi: efi_type
libzfs_core
libshare: nfs: is_shared_cb
libzpool
libzutil: import
libnvpair
libzfs
module: avl
module: zcommon: zprop
module: zfs: linux
module: zfs: freebsd
module: zfs
module: icp
```

**Comment by nabijaczleweli** [comment]
Squished. Loose version at https://github.com/nabijaczleweli/zfs/tree/lib-ARGSUSED-sparse if needed.

**Review by behlendorf — APPROVED** [review]
